### PR TITLE
ZK-4084: New added header is invisible when all headers have fixed width

### PR DESCRIPTION
**Pull request by b952426**
zk-4084: New added header is invisible when all headers have fixed width

### CODE REVIEW

**Review by scribetw — CHANGES_REQUESTED** [review]
Fix MeshWidget.js _fixHeaders

File: zul/src/archive/web/js/zul/mesh/MeshWidget.js [hunk 1]
```diff
@@ -800,6 +800,7 @@ zul.mesh.MeshWidget = zk.$extends(zul.Widget, {
 			if (this._cssflex && this.isChildrenFlex()) return;
 			var empty = true,
 				flex = false,
+				addHead = false,
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/mesh/MeshWidget.js:803]: shouldFix

File: zul/src/archive/web/js/zul/mesh/MeshWidget.js [hunk 2]
```diff
@@ -820,6 +821,9 @@ zul.mesh.MeshWidget = zk.$extends(zul.Widget, {
 					if (w.getLabel() || w.getImage() || w.nChildren) {
 						emptyHeader = false;
 						empty = false;
+						if (w.getWidth() == undefined) {
+							addHead = true;
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/mesh/MeshWidget.js:825]: check !frozen && ehead.style.width first

File: zul/src/archive/web/js/zul/mesh/MeshWidget.js [hunk 3]
```diff
@@ -843,7 +847,11 @@ zul.mesh.MeshWidget = zk.$extends(zul.Widget, {
 					delete w._nhflexbak;
 				}
 			}
-			return old != this.ehead.style.display;
+			if (addHead) {
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/mesh/MeshWidget.js:850]: return a || b

**Review by scribetw — CHANGES_REQUESTED** [review]

File: zktest/src/archive/test2/B95-ZK-4084.zul [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+		Description:
+
+		History:
+				Thu Dec 17 11:01:10 CST 2020, Created by katherinelin
+
+Copyright (C) 2020 Potix Corporation. All Rights Reserved.
+-->
+<zk>
+	<label multiline="true">
+		1. click the append button, should append the new listheader.
```
Comment (scribetw) [inline zktest/src/archive/test2/B95-ZK-4084.zul:16]: should see 3

File: zktest/test/java/org/zkoss/zktest/zats/test2/B95_ZK_4084Test.java [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+	@Test
+	public void test() {
+		connect();
+		click(jq("@button").eq(0));
+		waitResponse();
+		Assert.assertNotEquals(0, jq(".z-listheader").eq(2).width());
+		Assert.assertNotEquals(400, jq(".z-listhead").width());
+		click(jq("@button").eq(1));
+		waitResponse();
+		Assert.assertEquals(jq(".z-listhead").width(), 400);
```
Comment (scribetw) [inline zktest/test/java/org/zkoss/zktest/zats/test2/B95_ZK_4084Test.java:28]: assertEquals(400, jq(".z-listhead").width());

File: zul/src/archive/web/js/zul/mesh/MeshWidget.js [hunk 6]
```diff
@@ -800,6 +800,7 @@ zul.mesh.MeshWidget = zk.$extends(zul.Widget, {
 			if (this._cssflex && this.isChildrenFlex()) return;
 			var empty = true,
 				flex = false,
+				shouldFix = false,
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/mesh/MeshWidget.js:803]: shouldCheckEmptyWidth = !this.frozen && this.eheadtbl && this.eheadtbl.style.width

File: zul/src/archive/web/js/zul/mesh/MeshWidget.js [hunk 7]
```diff
@@ -820,6 +821,9 @@ zul.mesh.MeshWidget = zk.$extends(zul.Widget, {
 					if (w.getLabel() || w.getImage() || w.nChildren) {
 						emptyHeader = false;
 						empty = false;
+						if (!this.frozen && this.eheadtbl.style.width && w.getWidth() == undefined) {
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/mesh/MeshWidget.js:824]: if (shouldCheckEmptyWidth && w.getWidth())

**Review by scribetw — CHANGES_REQUESTED** [review]

**Review by plijyfes — APPROVED** [review]

**Review by scribetw — APPROVED** [review]

**Comment by scribetw** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes, enjoy!

File: zul/src/archive/web/js/zul/mesh/MeshWidget.js [hunk 8]
```diff
@@ -225,7 +225,7 @@ it will be useful, but WITHOUT ANY WARRANTY.
 			meshmin = wgt._hflex == 'min';
 		if (!wgt.head && (meshmin || sbc)) {
 			var bdw = zk(wgt.$n()).padBorderWidth(),
-				wd = _getMinWd(wgt) + bdw, // has to call _getMinWd first so wgt._minWd will be available
+				wd = _getMinWd(wgt) + (zk(wgt.ebody).hasVScroll() ? jq.scrollbarWidth() : 0) + bdw, // has to call _getMinWd first so wgt._minWd will be available
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/mesh/MeshWidget.js:228]:
```suggestion
				wd = _getMinWd(wgt) + bdw, // has to call _getMinWd first so wgt._minWd will be available
```

File: zul/src/archive/web/js/zul/mesh/MeshWidget.js [hunk 9]
```diff
@@ -820,6 +822,9 @@ zul.mesh.MeshWidget = zk.$extends(zul.Widget, {
 					if (w.getLabel() || w.getImage() || w.nChildren) {
 						emptyHeader = false;
 						empty = false;
+						if (shouldCheckEmptyWidth && w.getWidth()) {
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/mesh/MeshWidget.js:825]: shouldn't be nested

File: zktest/src/archive/test2/config.properties [hunk 10]
```diff
@@ -2897,6 +2897,7 @@ B90-ZK-4431.zul=A,E,Multislider
 ##zats##B95-ZK-4753.zul=A,E,Navbar,Nav,Navitem,focus,a11y
 ##zats##B95-ZK-4732.zul=A,E,Tree,client,ROD,scope
 ##zats##B95-ZK-4732-2.zul=A,E,Tree,client,ROD,scope
+##zats##B95-ZK-4084.zul=A,E,Width,AddChild
```
Comment (scribetw) [inline zktest/src/archive/test2/config.properties:2900]: Need MeshWidget,HeadWidget at least

**Review by scribetw — CHANGES_REQUESTED** [review]

**Review by scribetw — APPROVED** [review]

**Comment by scribetw** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes, enjoy!

**Pull request merged**